The no-ports end-to-end tests fail when executed with atSigns that are not production atSigns.

### Describe the bug

In "main.c",since null is passed to the atclient_utils_find_atserver_address() function in "atclient.c", the default values of production root server and port are used. Hence the test pass. However, during local testing, if options are provided to the monitor_pkam_authenticate method (the last parameter in method), the atdirectory_host and atdirectory_port, corresponding to the root server and port, are assigned to atserver_host and atserver_port. This leads to the test failure.

Additionally, the "atclient_utils_find_atserver_address" method always sets the root server details to the production host and port. Any alternative root server and port provided through the atclient_pkam_authenticate_options are ignored.

noports/packages/c/sshnpd/src/main.c
```
res = atclient_monitor_pkam_authenticate(&monitor_ctx, params.atsign, &atkeys, NULL);
```

packages/atclient/src/atclient.c-> atclient_pkam_authenticate()
```
if (options != NULL) {
    if (atclient_pkam_authenticate_options_is_atdirectory_host_initialized(options) &&
        atclient_pkam_authenticate_options_is_atdirectory_port_initialized(options)) {
      atserver_host = options->atdirectory_host;
      atserver_port = options->atdirectory_port;
    }
  }

  if (atserver_host == NULL || atserver_port == 0) {
    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_INFO,
                 "Missing atServer host or port. Using production atDirectory to look up atServer host and port\n");
    if ((ret = atclient_utils_find_atserver_address(ATCLIENT_ATDIRECTORY_PRODUCTION_HOST,
                                                    ATCLIENT_ATDIRECTORY_PRODUCTION_PORT, atsign, &atserver_host,
                                                    &atserver_port)) != 0) {
      atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_utils_find_atserver_address: %d\n", ret);
      goto exit;
    }
    // only free this memory if it was allocated internally (by atclient_utils_find_atserver_address)
    should_free_atserver_host = true;
  }
```

Log snippet of test failure:

```
Verbose mode enabled
[DEBG] 2024-09-30 13:45:35.769157 | sshnpd | Using atkeysfile: /home/sitaram/.atsign/keys/@bob🛠_key.atKeys
[DEBG] 2024-09-30 13:45:35.838057 | connection | 	SENT: "bob🛠"
[DEBG] 2024-09-30 13:45:35.838395 | connection | 	RECV: "vip.ve.atsign.zone:25003"
[INFO] 2024-09-30 13:45:35.838444 | atclient | Missing atServer host or port. Using production atDirectory to look up atServer host and port
[DEBG] 2024-09-30 13:45:37.368938 | connection | 	SENT: "bob🛠"
[DEBG] 2024-09-30 13:45:37.675274 | connection | 	RECV: "null"
[ERROR] 2024-09-30 13:45:37.675314 | atclient_utils | Failed to parse port from atdirectory response
```

### Steps to reproduce

Run no-ports e2e tests with local atSigns.

Steps to run noports e2e tests locally:

In a NoPorts repo clone

    - build the binaries (cd /path/to/noports/packages/dart/sshnoports; buildBinaries)
    - run an srvd
        - bin/srvd -a @chuck --root-domain vip.ve.atsign.zone --ip 127.0.0.1 -v
    - run the e2e tests
         - /path/to/noports_repo/tests/e2e_all/scripts/main.sh @alice @bob @chuck -w 5 -r vip.ve.atsign.zone

### Expected behavior

For the "atclient_pkam_authenticate", if null is passed for "atclient_pkam_authenticate_options" then the default values should be considered. If options are set, the root server and port from the options should be used to fetch the secondary server address.

--- a/examples/desktop/repl/src/main.c
+++ b/examples/desktop/repl/src/main.c
@@ -133,12 +133,12 @@ static int set_up_pkam_auth_options(atclient_authenticate_options *pkam_authenti
       goto exit;
     }
 
-    if ((ret = atclient_authenticate_options_set_at_directory_host(pkam_authenticate_options, root_host)) != 0) {
+    if ((ret = atclient_authenticate_options_set_atdirectory_host(pkam_authenticate_options, root_host)) != 0) {
       atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to set at directory host\n");
       goto exit;
     }
 
-    if ((ret = atclient_authenticate_options_set_at_directory_port(pkam_authenticate_options, root_port)) != 0) {
+    if ((ret = atclient_authenticate_options_set_atdirectory_port(pkam_authenticate_options, root_port)) != 0) {
       atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to set at directory port\n");
       goto exit;
     }

--- a/packages/atclient/include/atclient/request_options.h
+++ b/packages/atclient/include/atclient/request_options.h
@@ -39,12 +39,12 @@ extern "C" {
 #define ATCLIENT_GET_ATKEYS_REQUEST_OPTIONS_REGEX_INITIALIZED (VALUE_INITIALIZED << 0)
 #define ATCLIENT_GET_ATKEYS_REQUEST_OPTIONS_SHOW_HIDDEN_INITIALIZED (VALUE_INITIALIZED << 1)
 
-#define ATCLIENT_AUTHENTICATE_OPTIONS_AT_DIRECTORY_HOST_INDEX 0
-#define ATCLIENT_AUTHENTICATE_OPTIONS_AT_DIRECTORY_PORT_INDEX 0
+#define ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_HOST_INDEX 0
+#define ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_PORT_INDEX 0
 #define ATCLIENT_AUTHENTICATE_OPTIONS_ATSERVER_HOST_INDEX 0
 #define ATCLIENT_AUTHENTICATE_OPTIONS_ATSERVER_PORT_INDEX 0
-#define ATCLIENT_AUTHENTICATE_OPTIONS_AT_DIRECTORY_HOST_INITIALIZED (VALUE_INITIALIZED << 0)
-#define ATCLIENT_AUTHENTICATE_OPTIONS_AT_DIRECTORY_PORT_INITIALIZED (VALUE_INITIALIZED << 1)
+#define ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_HOST_INITIALIZED (VALUE_INITIALIZED << 0)
+#define ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_PORT_INITIALIZED (VALUE_INITIALIZED << 1)
 #define ATCLIENT_AUTHENTICATE_OPTIONS_ATSERVER_HOST_INITIALIZED (VALUE_INITIALIZED << 2)
 #define ATCLIENT_AUTHENTICATE_OPTIONS_ATSERVER_PORT_INITIALIZED (VALUE_INITIALIZED << 3)
 
@@ -260,12 +260,12 @@ void atclient_authenticate_options_init(atclient_authenticate_options *options);
 void atclient_authenticate_options_free(atclient_authenticate_options *options);
 
 bool atclient_authenticate_options_is_atdirectory_host_initialized(const atclient_authenticate_options *options);
-int atclient_authenticate_options_set_at_directory_host(atclient_authenticate_options *options, char *atdirectory_host);
-void atclient_authenticate_options_unset_at_directory_host(atclient_authenticate_options *options);
+int atclient_authenticate_options_set_atdirectory_host(atclient_authenticate_options *options, char *atdirectory_host);
+void atclient_authenticate_options_unset_atdirectory_host(atclient_authenticate_options *options);
 
 bool atclient_authenticate_options_is_atdirectory_port_initialized(const atclient_authenticate_options *options);
-int atclient_authenticate_options_set_at_directory_port(atclient_authenticate_options *options, int atdirectory_port);
-void atclient_authenticate_options_unset_at_directory_port(atclient_authenticate_options *options);
+int atclient_authenticate_options_set_atdirectory_port(atclient_authenticate_options *options, int atdirectory_port);
+void atclient_authenticate_options_unset_atdirectory_port(atclient_authenticate_options *options);
 
 bool atclient_authenticate_options_is_atserver_host_initialized(const atclient_authenticate_options *options);
 int atclient_authenticate_options_set_atserver_host(atclient_authenticate_options *options, char *atserver_host);

--- a/packages/atclient/src/atclient.c
+++ b/packages/atclient/src/atclient.c
@@ -203,12 +203,8 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
   /*
    * 2. Initialize variables
    */
-
-  // free later
-  char *root_cmd = NULL;
-  char *from_cmd = NULL;
-  char *pkam_cmd = NULL;
-  char *atsign_with_at = NULL;
+  char *atdirectory_host = NULL;
+  int atdirectory_port = 0;
 
   char *atserver_host = NULL;
   int atserver_port = 0;
@@ -233,6 +229,12 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
   memset(signature_base64, 0, sizeof(unsigned char) * signature_base64_size);
   size_t signature_base64_len = 0;
 
+  // free later
+  char *root_cmd = NULL;
+  char *from_cmd = NULL;
+  char *pkam_cmd = NULL;
+  char *atsign_with_at = NULL;
+
   /*
    * 3. Ensure that the atsign has the @ symbol.
    */
@@ -244,22 +246,34 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
   const char *atsign_without_at = atsign_with_at + 1;
 
   /*
-   * 4. Get atserver_host and atserver_port
+   * 4. Get atdirectory_host and atdirectory_port
    */
-  bool should_free_atserver_host = false;
-  if (options != NULL) {
-    if (atclient_authenticate_options_is_atdirectory_host_initialized(options) &&
-        atclient_authenticate_options_is_atdirectory_port_initialized(options)) {
-      atserver_host = options->atdirectory_host;
-      atserver_port = options->atdirectory_port;
-    }
+  if(options != NULL && atclient_authenticate_options_is_atdirectory_host_initialized(options) && options->atdirectory_host != NULL &&
+     atclient_authenticate_options_is_atdirectory_port_initialized(options) && options->atdirectory_port != 0) {
+    atdirectory_host = options->atdirectory_host;
+    atdirectory_port = options->atdirectory_port;
+  } else {
+    atdirectory_host = ATCLIENT_ATDIRECTORY_PRODUCTION_HOST;
+    atdirectory_port = ATCLIENT_ATDIRECTORY_PRODUCTION_PORT;
+  }
+
+  /*
+   * 5. Get atserver_host and atserver_port
+   */
+  bool should_free_atserver_host;
+  if (options != NULL && atclient_authenticate_options_is_atserver_host_initialized(options) &&
+      options->atserver_host != NULL && atclient_authenticate_options_is_atserver_port_initialized(options) &&
+      options->atserver_port != 0) {
+    atserver_host = options->atserver_host;
+    atserver_port = options->atserver_port;
+    should_free_atserver_host = true;
   }
 
   if (atserver_host == NULL || atserver_port == 0) {
     atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_INFO,
-                 "Missing atServer host or port. Using production atDirectory to look up atServer host and port\n");
-    if ((ret = atclient_utils_find_atserver_address(ATCLIENT_ATDIRECTORY_PRODUCTION_HOST,
-                                                    ATCLIENT_ATDIRECTORY_PRODUCTION_PORT, atsign, &atserver_host,
+                 "Missing atServer host or port. Using %s:%lu atDirectory to find atServer address\n", atdirectory_host, atdirectory_port);
+    if ((ret = atclient_utils_find_atserver_address(atdirectory_host,
+                                                    atdirectory_port, atsign, &atserver_host,
                                                     &atserver_port)) != 0) {
       atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_utils_find_atserver_address: %d\n", ret);
       goto exit;
@@ -269,7 +283,7 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
   }
 
   /*
-   * 5. Start atServer connection (kill the existing connection if it exists)
+   * 6. Start atServer connection (kill the existing connection if it exists)
    */
   if ((ret = atclient_start_atserver_connection(ctx, atserver_host, atserver_port)) != 0) {
     atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_start_atserver_connection: %d\n", ret);
@@ -277,7 +291,7 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
   }
 
   /*
-   * 6a. Build `from:` cmd
+   * 7a. Build `from:` cmd
    */
   const size_t from_cmd_size =
       strlen("from:") + strlen(atsign_without_at) + strlen("\r\n") + 1; // "from:" has a length of 5
@@ -288,7 +302,7 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
   snprintf(from_cmd, from_cmd_size, "from:%s\r\n", atsign_without_at);
 
   /*
-   * 6b. Send `from:` cmd
+   * 7b. Send `from:` cmd
    */
   if ((ret = atclient_connection_send(&(ctx->atserver_connection), (unsigned char *)from_cmd, from_cmd_size - 1, recv,
                                       recvsize, &recv_len)) != 0) {
@@ -305,7 +319,7 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
   }
 
   /*
-   * 7. We got `data:<challenge>`
+   * 8. We got `data:<challenge>`
    *    Let us sign the challenge with RSA-2048 PKAM Private Key and Base64 Encode it
    */
 
@@ -327,7 +341,7 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
   }
 
   /*
-   * 8a. Build `pkam:` noop_cmd
+   * 9a. Build `pkam:` noop_cmd
    */
   size_t pkam_cmd_size = strlen("pkam:");
   if (atclient_atkeys_is_enrollment_id_initialized((atclient_atkeys *)atkeys) && atkeys->enrollment_id != NULL) {
@@ -348,7 +362,7 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
   pos += snprintf(pkam_cmd + pos, pkam_cmd_size - pos, "%s\r\n", signature_base64);
 
   /*
-   * 8b. Send `pkam:` noop_cmd
+   * 9b. Send `pkam:` noop_cmd
    */
   memset(recv, 0, sizeof(unsigned char) * recvsize);
   if ((ret = atclient_connection_send(&(ctx->atserver_connection), (unsigned char *)pkam_cmd, pkam_cmd_size - 1, recv,
@@ -366,7 +380,7 @@ int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient
   }
 
   /*
-   * 9. Set up the atclient context
+   * 10. Set up the atclient context
    */
 
   // initialize ctx->atsign.atsign and ctx->atsign.withour_prefix_str to the newly authenticated atSign

--- a/packages/atclient/src/request_options.c
+++ b/packages/atclient/src/request_options.c
@@ -9,16 +9,34 @@
 
 #define TAG "request_options"
 
-// default param values for ATCLIENT PKAM AUTHENTICATE OPTIONS
-#define ATCLIENT_AUTHENTICATE_DEFAULT_ROOT_SERVER_HOST "root.atsign.org"
-#define ATCLIENT_AUTHENTICATE_DEFAULT_ROOT_SERVER_PORT 64
+/*
+ * 1A. Put Self Key
+ */
+static void atclient_get_self_key_request_options_set_store_atkey_metadata_initialized(
+    atclient_get_self_key_request_options *options, const bool initialized);
 
+/*
+ * 1B. Put SharedKey
+ */
 static void atclient_put_shared_key_request_options_set_shared_encryption_key_initialized(
     atclient_put_shared_key_request_options *options, const bool initialized);
 static void atclient_put_shared_key_request_options_set_iv_initialized(atclient_put_shared_key_request_options *options,
                                                                        const bool initialized);
+
+/*
+ * 1C. Put PublicKey
+ */
+// empty for now
+
+/*
+ * 2A. Get SelfKey
+ */
 static void atclient_get_self_key_request_options_set_store_atkey_metadata_initialized(
     atclient_get_self_key_request_options *options, const bool initialized);
+
+/*
+ * 2B. Get SharedKey
+ */
 static void atclient_get_shared_key_request_options_set_shared_encryption_key_initialized(
     atclient_get_shared_key_request_options *options, const bool initialized);
 static void atclient_get_shared_key_request_options_set_iv_initialized(atclient_get_shared_key_request_options *options,
@@ -28,19 +46,49 @@ atclient_get_shared_key_request_options_set_bypass_cache_initialized(atclient_ge
                                                                      const bool initialized);
 static void atclient_get_shared_key_request_options_set_store_atkey_metadata_initialized(
     atclient_get_shared_key_request_options *options, const bool initialized);
+
+/*
+ * 2C. Get PublicKey
+ */
 static void
 atclient_get_public_key_request_options_set_bypass_cache_initialized(atclient_get_public_key_request_options *options,
                                                                      const bool initialized);
 static void atclient_get_public_key_request_options_set_store_atkey_metadata_initialized(
     atclient_get_public_key_request_options *options, const bool initialized);
+
+/*
+ * 3. Delete
+ */
+static void
+atclient_delete_request_options_set_skip_shared_by_check_intitialized(atclient_delete_request_options *options,
+                                                                      const bool initialized);
+
+/*
+ * 4. Get AtKeys
+ */
 static void atclient_get_atkeys_request_options_set_regex_initialized(atclient_get_atkeys_request_options *options,
                                                                       const bool initialized);
 static void
 atclient_get_atkeys_request_options_set_show_hidden_initialized(atclient_get_atkeys_request_options *options,
                                                                 const bool initialized);
+
+/*
+ * 5. Authenticate Request Options
+ */
 static void
-atclient_delete_request_options_skip_shared_by_check_set_intitialized(atclient_delete_request_options *options,
-                                                                      const bool initialized);
+atclient_authenticate_options_set_atdirectory_host_initialized(atclient_authenticate_options *options,
+                                                                       const bool initialized);
+
+static void
+atclient_authenticate_options_set_atdirectory_port_initialized(atclient_authenticate_options *options,
+                                                                       const bool initialized);
+
+static void atclient_authenticate_options_set_atserver_host_initialized(atclient_authenticate_options *options,
+                                                                                const bool initialized);
+
+static void atclient_authenticate_options_set_atserver_port_initialized(atclient_authenticate_options *options,
+                                                                                const bool initialized);
+
 /*
  * =================
  * 1A. Put SelfKey
@@ -1226,8 +1274,8 @@ void atclient_delete_request_options_free(atclient_delete_request_options *optio
   }
 
   /*
-  * 2. Reset the value
-  */
+   * 2. Reset the value
+   */
   if (atclient_delete_request_options_is_skip_shared_by_check_flag_initialized(options)) {
     atclient_delete_request_options_unset_skip_shared_by_check(options);
   }
@@ -1252,7 +1300,7 @@ bool atclient_delete_request_options_is_skip_shared_by_check_flag_initialized(
 }
 
 static void
-atclient_delete_request_options_skip_shared_by_check_set_intitialized(atclient_delete_request_options *options,
+atclient_delete_request_options_set_skip_shared_by_check_intitialized(atclient_delete_request_options *options,
                                                                       const bool initialized) {
   /*
    * 1. Validate arguments
@@ -1296,8 +1344,8 @@ void atclient_delete_request_options_set_skip_shared_by_check(atclient_delete_re
   /*
    * 3. Set value and set intiliazed to true
    */
-   memcpy(&(options->skip_shared_by_check), &option, sizeof(option));
-  atclient_delete_request_options_skip_shared_by_check_set_intitialized(options, true);
+  memcpy(&(options->skip_shared_by_check), &option, sizeof(option));
+  atclient_delete_request_options_set_skip_shared_by_check_intitialized(options, true);
 }
 
 void atclient_delete_request_options_unset_skip_shared_by_check(atclient_delete_request_options *options) {
@@ -1311,7 +1359,7 @@ void atclient_delete_request_options_unset_skip_shared_by_check(atclient_delete_
   }
 
   options->skip_shared_by_check = false;
-  atclient_delete_request_options_skip_shared_by_check_set_intitialized(options, false);
+  atclient_delete_request_options_set_skip_shared_by_check_intitialized(options, false);
 }
 
 void atclient_get_atkeys_request_options_init(atclient_get_atkeys_request_options *options) {
@@ -1551,8 +1599,8 @@ void atclient_get_atkeys_request_options_unset_show_hidden(atclient_get_atkeys_r
 
 void atclient_authenticate_options_init(atclient_authenticate_options *options) {
   memset(options, 0, sizeof(atclient_authenticate_options));
-  options->atdirectory_host = ATCLIENT_AUTHENTICATE_DEFAULT_ROOT_SERVER_HOST;
-  options->atdirectory_port = ATCLIENT_AUTHENTICATE_DEFAULT_ROOT_SERVER_PORT;
+  options->atdirectory_host = ATCLIENT_ATDIRECTORY_PRODUCTION_HOST;
+  options->atdirectory_port = ATCLIENT_ATDIRECTORY_PRODUCTION_PORT;
   options->atserver_host = NULL;
   options->atserver_port = 0;
 }
@@ -1571,11 +1619,11 @@ void atclient_authenticate_options_free(atclient_authenticate_options *options) 
    */
 
   if (atclient_authenticate_options_is_atdirectory_host_initialized(options)) {
-    atclient_authenticate_options_unset_at_directory_host(options);
+    atclient_authenticate_options_unset_atdirectory_host(options);
   }
 
   if (atclient_authenticate_options_is_atdirectory_port_initialized(options)) {
-    atclient_authenticate_options_unset_at_directory_port(options);
+    atclient_authenticate_options_unset_atdirectory_port(options);
   }
 
   if (atclient_authenticate_options_is_atserver_host_initialized(options)) {
@@ -1601,8 +1649,8 @@ bool atclient_authenticate_options_is_atdirectory_host_initialized(const atclien
   /*
    * 2. Check if the at directory host is initialized
    */
-  return options->_initialized_fields[ATCLIENT_AUTHENTICATE_OPTIONS_AT_DIRECTORY_HOST_INDEX] &
-         ATCLIENT_AUTHENTICATE_OPTIONS_AT_DIRECTORY_HOST_INITIALIZED;
+  return options->_initialized_fields[ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_HOST_INDEX] &
+         ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_HOST_INITIALIZED;
 }
 
 bool atclient_authenticate_options_is_atdirectory_port_initialized(const atclient_authenticate_options *options) {
@@ -1619,8 +1667,8 @@ bool atclient_authenticate_options_is_atdirectory_port_initialized(const atclien
   /*
    * Check if the at directory port is initialized
    */
-  return options->_initialized_fields[ATCLIENT_AUTHENTICATE_OPTIONS_AT_DIRECTORY_PORT_INDEX] &
-         ATCLIENT_AUTHENTICATE_OPTIONS_AT_DIRECTORY_PORT_INITIALIZED;
+  return options->_initialized_fields[ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_PORT_INDEX] &
+         ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_PORT_INITIALIZED;
 }
 
 bool atclient_authenticate_options_is_atserver_host_initialized(const atclient_authenticate_options *options) {
@@ -1659,34 +1707,36 @@ bool atclient_authenticate_options_is_atserver_port_initialized(const atclient_a
          ATCLIENT_AUTHENTICATE_OPTIONS_ATSERVER_PORT_INITIALIZED;
 }
 
-void atclient_authenticate_options_unset_at_directory_host(atclient_authenticate_options *options) {
+void atclient_authenticate_options_unset_atdirectory_host(atclient_authenticate_options *options) {
   /*
    * 1. Validate arguments
    */
   if (options == NULL) {
     atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR,
-                 "atclient_authenticate_options_unset_at_directory_host: Invalid arguments\n");
+                 "atclient_authenticate_options_unset_atdirectory_host: Invalid arguments\n");
     return;
   }
 
   /*
    * 2. Unsetat directory   */
-  if (atclient_authenticate_options_is_atdirectory_host_initialized(options)) {
+  if (atclient_authenticate_options_is_atdirectory_host_initialized(options) && options->atdirectory_host != NULL) {
     free(options->atdirectory_host);
   }
   options->atdirectory_host = NULL;
+  atclient_authenticate_options_set_atdirectory_host_initialized(options, false);
 }
 
-void atclient_authenticate_options_unset_at_directory_port(atclient_authenticate_options *options) {
+void atclient_authenticate_options_unset_atdirectory_port(atclient_authenticate_options *options) {
   /*
    * 1. Validate arguments
    */
   if (options == NULL) {
     atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR,
-                 "atclient_authenticate_options_unset_at_directory_port: Invalid arguments\n");
+                 "atclient_authenticate_options_unset_atdirectory_port: Invalid arguments\n");
     return;
   }
-  options->atdirectory_port = 64;
+  options->atdirectory_port = ATCLIENT_ATDIRECTORY_PRODUCTION_PORT;
+  atclient_authenticate_options_set_atdirectory_port_initialized(options, false);
 }
 
 void atclient_authenticate_options_unset_atserver_host(atclient_authenticate_options *options) {
@@ -1702,10 +1752,11 @@ void atclient_authenticate_options_unset_atserver_host(atclient_authenticate_opt
   /*
    * 2. Unset atserver host
    */
-  if (atclient_authenticate_options_is_atserver_host_initialized(options)) {
+  if (atclient_authenticate_options_is_atserver_host_initialized(options) && options->atserver_host != NULL) {
     free(options->atserver_host);
   }
   options->atserver_host = NULL;
+  atclient_authenticate_options_set_atserver_host_initialized(options, false);
 }
 
 void atclient_authenticate_options_unset_atserver_port(atclient_authenticate_options *options) {
@@ -1718,9 +1769,10 @@ void atclient_authenticate_options_unset_atserver_port(atclient_authenticate_opt
     return;
   }
   options->atserver_port = 0;
+  atclient_authenticate_options_set_atserver_port_initialized(options, false);
 }
 
-int atclient_authenticate_options_set_at_directory_host(atclient_authenticate_options *options,
+int atclient_authenticate_options_set_atdirectory_host(atclient_authenticate_options *options,
                                                         char *atdirectory_host) {
   int ret = 1;
 
@@ -1730,7 +1782,7 @@ int atclient_authenticate_options_set_at_directory_host(atclient_authenticate_op
   if (options == NULL) {
     ret = 1;
     atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR,
-                 "atclient_authenticate_options_set_at_directory_host: Invalid arguments\n");
+                 "atclient_authenticate_options_set_atdirectory_host: Invalid arguments\n");
     return ret;
   }
 
@@ -1738,19 +1790,34 @@ int atclient_authenticate_options_set_at_directory_host(atclient_authenticate_op
    * 2. Unset at directory host, if initialized
    */
   if (atclient_authenticate_options_is_atdirectory_host_initialized(options)) {
-    atclient_authenticate_options_unset_at_directory_host(options);
+    atclient_authenticate_options_unset_atdirectory_host(options);
   }
 
   /*
    * 3. Set at directory host
    */
-  options->atdirectory_host = atdirectory_host;
+  const size_t atdirectory_host_size = strlen(atdirectory_host) + 1;
+  if ((options->atdirectory_host = (char *)malloc(sizeof(char) * atdirectory_host_size)) == NULL) {
+    ret = 1;
+    atlogger_log(
+        TAG, ATLOGGER_LOGGING_LEVEL_ERROR,
+        "atclient_authenticate_options_set_atdirectory_host: Failed to allocate memory for at directory host\n");
+    goto exit;
+  }
+
+  atclient_authenticate_options_set_atdirectory_host_initialized(options, true);
+
+  const size_t atdirectory_host_len = strlen(atdirectory_host);
+  memcpy(options->atdirectory_host, atdirectory_host, atdirectory_host_len);
+  options->atdirectory_host[atdirectory_host_len] = '\0';
+
+
   ret = 0;
   goto exit;
 exit: { return ret; }
 }
 
-int atclient_authenticate_options_set_at_directory_port(atclient_authenticate_options *options, int atdirectory_port) {
+int atclient_authenticate_options_set_atdirectory_port(atclient_authenticate_options *options, int atdirectory_port) {
   int ret = 1;
 
   /*
@@ -1759,7 +1826,7 @@ int atclient_authenticate_options_set_at_directory_port(atclient_authenticate_op
   if (options == NULL) {
     ret = 1;
     atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR,
-                 "atclient_authenticate_options_set_at_directory_port: Invalid arguments\n");
+                 "atclient_authenticate_options_set_atdirectory_port: Invalid arguments\n");
     return ret;
   }
 
@@ -1767,13 +1834,14 @@ int atclient_authenticate_options_set_at_directory_port(atclient_authenticate_op
    * 2. Unset at directory port, if initialized
    */
   if (atclient_authenticate_options_is_atdirectory_port_initialized(options)) {
-    atclient_authenticate_options_unset_at_directory_host(options);
+    atclient_authenticate_options_unset_atdirectory_host(options);
   }
 
   /*
    * 3. Set at directory port
    */
   options->atdirectory_port = atdirectory_port;
+  atclient_authenticate_options_set_atdirectory_port_initialized(options, true);
   ret = 0;
   goto exit;
 exit: { return ret; }
@@ -1802,7 +1870,19 @@ int atclient_authenticate_options_set_atserver_host(atclient_authenticate_option
   /*
    * 3. Set atserver host
    */
-  options->atserver_host = atserver_host;
+  const size_t atserver_host_size = strlen(atserver_host) + 1;
+  if((options->atserver_host = (char *)malloc(sizeof(char) * atserver_host_size)) == NULL) {
+    ret = 1;
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR,
+                 "atclient_authenticate_options_set_atserver_host: Failed to allocate memory for atserver host\n");
+    return ret;
+  }
+
+  const size_t atserver_host_len = strlen(atserver_host);
+  memcpy(options->atserver_host, atserver_host, atserver_host_len);
+  options->atserver_host[atserver_host_len] = '\0';
+
+  atclient_authenticate_options_set_atserver_host_initialized(options, true);
 
   ret = 0;
   goto exit;
@@ -1834,9 +1914,102 @@ int atclient_authenticate_options_set_atserver_port(atclient_authenticate_option
    * 3. Set atserver port
    */
   options->atserver_port = atserver_port;
+  atclient_authenticate_options_set_atserver_port_initialized(options, true);
 
   ret = 0;
   goto exit;
 
 exit: { return ret; }
+}
+
+static void atclient_authenticate_options_set_atdirectory_host_initialized(atclient_authenticate_options *options,
+                                                                           const bool initialized) {
+  /*
+   * 1. Validate arguments
+   */
+  if (options == NULL) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_authenticate_options_set_atdirectory_host_initialized: "
+                                                    "Invalid arguments\n");
+    return;
+  }
+
+  /*
+   * 2. Set the at directory host initialized
+   */
+  if (initialized) {
+    options->_initialized_fields[ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_HOST_INDEX] |=
+        ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_HOST_INITIALIZED;
+  } else {
+    options->_initialized_fields[ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_HOST_INDEX] &=
+        ~ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_HOST_INITIALIZED;
+  }
+}
+
+static void atclient_authenticate_options_set_atdirectory_port_initialized(atclient_authenticate_options *options,
+                                                                           const bool initialized) {
+  /*
+   * 1. Validate arguments
+   */
+  if (options == NULL) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_authenticate_options_set_atdirectory_port_initialized: "
+                                                    "Invalid arguments\n");
+    return;
+  }
+
+  /*
+   * 2. Set the at directory port initialized
+   */
+  if (initialized) {
+    options->_initialized_fields[ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_PORT_INDEX] |=
+        ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_PORT_INITIALIZED;
+  } else {
+    options->_initialized_fields[ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_PORT_INDEX] &=
+        ~ATCLIENT_AUTHENTICATE_OPTIONS_ATDIRECTORY_PORT_INITIALIZED;
+  }
+}
+
+static void atclient_authenticate_options_set_atserver_host_initialized(atclient_authenticate_options *options,
+                                                                        const bool initialized) {
+  /*
+   * 1. Validate arguments
+   */
+  if (options == NULL) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_authenticate_options_set_atserver_host_initialized: "
+                                                    "Invalid arguments\n");
+    return;
+  }
+
+  /*
+   * 2. Set the atserver host initialized
+   */
+  if (initialized) {
+    options->_initialized_fields[ATCLIENT_AUTHENTICATE_OPTIONS_ATSERVER_HOST_INDEX] |=
+        ATCLIENT_AUTHENTICATE_OPTIONS_ATSERVER_HOST_INITIALIZED;
+  } else {
+    options->_initialized_fields[ATCLIENT_AUTHENTICATE_OPTIONS_ATSERVER_HOST_INDEX] &=
+        ~ATCLIENT_AUTHENTICATE_OPTIONS_ATSERVER_HOST_INITIALIZED;
+  }
+}
+
+static void atclient_authenticate_options_set_atserver_port_initialized(atclient_authenticate_options *options,
+                                                                        const bool initialized) {
+  /*
+   * 1. Validate arguments
+   */
+  if (options == NULL) {
+    atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_authenticate_options_set_atserver_port_initialized: "
+                                                    "Invalid arguments\n");
+    return;
+  }
+
+  /*
+   * 2. Set the atserver port initialized
+   */
+  if (initialized) {
+    options->_initialized_fields[ATCLIENT_AUTHENTICATE_OPTIONS_ATSERVER_PORT_INDEX] |=
+        ATCLIENT_AUTHENTICATE_OPTIONS_ATSERVER_PORT_INITIALIZED;
+  } else {
+    options->_initialized_fields[ATCLIENT_AUTHENTICATE_OPTIONS_ATSERVER_PORT_INDEX] &=
+        ~ATCLIENT_AUTHENTICATE_OPTIONS_ATSERVER_PORT_INITIALIZED;
+  }
 }

